### PR TITLE
Error overhaul

### PR DESCRIPTION
**Pull request by vegaro**
Easier way to detect if user cancelled purchase
---
- Adds `userCancelled` as a parameter to the completion block of the makePurchase function

Better error codes
---
- Creates `errors.kt` that contain `PurchasesErrorCode` and `BackendErrorCode`
- Updates `PurchasesError` to contain code,  message and an optional underlying error message
- Removes the old error domains.
- If there is a BillingClient error it maps the error code to our own set of errors and adds the error description as an underlying error message
- For backend errors, it maps the error code to our own errors and adds the backend response as the originating error message.
- Coming really hot: the entitlements call now fails if the service returns FEATURE_NOT_SUPPORTED or BILLING_UNAVAILABLE. Also that DEVELOPER_ERROR that was always displaying in the logs, now shows as "Billing Service Setup is already trying to connect."

Small refactor in `Backend.kt`
---
- Similar to iOS, I wanted to communicate from the Backend to the Purchases class if a purchase should be consumed and have the logic for it in the Backend class. I changed the error function of the `postReceiptData` to return a `shouldConsumePurchase: Boolean` along with the `PurchasesError`. That means the inner class `PurchaserInfoReceivingCall` in the Backend doesn't make sense anymore, so I removed it and moved its content to the `getPurchaserInfo` method. I also removed some duplicated code and created a `MutableMap.addCallback` extension function that checks if a call is already in progress and adds the callback, it's similar to `addCallback` in  `RCBackend.m` in the iOS SDK.

Important fix
---
Runnables were being submitted to the executor instead of executed. This makes any Exception that occurs in the Runnable to be swallowed and wrapped in a Future, that the `submit` call returns. Since we are not doing anything with the Future object, we are missing any potential exceptions that happen in the Runnable (i.e. there was a NullPointerException when the phone is in airplane mode)

### CODE REVIEW

File: purchases/src/main/kotlin/com/revenuecat/purchases/HTTPClient.kt [hunk 1]
```diff
@@ -93,34 +94,24 @@ internal class HTTPClient(
...
         } finally {
             connection.disconnect()
         }
 
-        try {
-            result.body = JSONObject(payload)
-            debugLog("${connection.requestMethod} $path ${result.responseCode}")
-        } catch (e: JSONException) {
-            log("Error parsing JSON ${e.localizedMessage}")
-            log("Data received: $payload")
```
Comment (vegaro) [inline purchases/src/main/kotlin/com/revenuecat/purchases/HTTPClient.kt]: might be good to log this somehow

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt [hunk 2]
```diff
@@ -449,8 +440,21 @@ class Purchases @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE) intern
...
+                        val missingProducts= populateSkuDetails(offerings, detailsByID)
+                        if (missingProducts.isNotEmpty()) {
+                            log("Could not find SkuDetails for ${missingProducts.joinToString(", ")}")
+                            log("Ensure your products are correctly configured in Play Store Developer Console")
+                        }
+                        if (offerings.mapNotNull { it.skuDetails }.isNotEmpty()) {
+                            cachedEntitlements = entitlements
+                        } else {
+                            cachedEntitlements = null
+                        }
```
Comment (vegaro) [inline purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt]: if it can't find any skudetails, it will retry next time, that's why this change. it was saving as cachedEntitlements sku details that could be null

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt [hunk 3]
```diff
@@ -559,8 +556,8 @@ class Purchases @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE) intern
                     cachePurchaserInfo(info)
                     sendUpdatedPurchaserInfoToDelegateIfChanged(info)
                     onSuccess(purchase, info)
-                }, { error ->
-                    if (error.code < HTTP_SERVER_ERROR) {
+                }, { error, shouldConsumePurchase ->
```
Comment (vegaro) [inline purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt]: moved this logic to the backend.kt

File: purchases/src/main/kotlin/com/revenuecat/purchases/Backend.kt [hunk 4]
```diff
@@ -119,30 +89,49 @@ internal class Backend(
...
-        synchronized(callbacks) {
-            if (!callbacks.containsKey(cacheKey)) {
-                callbacks[cacheKey] = mutableListOf(onSuccess to onError)
 
-                enqueue(call)
-            } else {
-                callbacks[cacheKey]!!.add(onSuccess to onError)
+            override fun onCompletion(result: HTTPClient.Result) {
+                synchronized(this) {
+                    postReceiptCallbacks.remove(cacheKey)?.forEach { (onSuccess, onError) ->
```
Comment (jeiting) [inline purchases/src/main/kotlin/com/revenuecat/purchases/Backend.kt]: very explicit. me gusta

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt [hunk 5]
```diff
@@ -449,8 +440,21 @@ class Purchases @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE) intern
             appUserID,
             { entitlements ->
                 getSkuDetails(entitlements, { detailsByID ->
-                    cachedEntitlements = entitlements
-                    populateSkuDetailsAndCallCompletion(detailsByID, entitlements, completion)
+                    entitlements.values.flatMap { it.offerings.values }.let { offerings ->
+                        val missingProducts= populateSkuDetails(offerings, detailsByID)
```
Comment (jeiting) [inline purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt]: little syntax issue with the equals

File: purchases/src/main/kotlin/com/revenuecat/purchases/errors.kt [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+enum class PurchasesErrorCode(val description: String) {
+    UnknownError("Unknown error."),
+    PurchaseCancelledError("Purchase was cancelled."),
+    StoreProblemError("There was a problem with the Play Store."),
+    PurchaseNotAllowedError("The device or user is not allowed to make the purchase."),
+    PurchaseInvalidError("One or more of the arguments provided are invalid."),
+    ProductNotAvailableForPurchaseError("The product is not available for purchase."),
+    ProductAlreadyPurchasedError("This product is already active for the user."),
```
Comment (jeiting) [inline purchases/src/main/kotlin/com/revenuecat/purchases/errors.kt]: Perhaps we should not treat this as an error? On iOS 12.2 this is actually treated as a health case,

Comment (vegaro) [reply purchases/src/main/kotlin/com/revenuecat/purchases/errors.kt]: Good idea, I am going to avoid return error and treat is a success, might help with some edge cases too

Comment (jeiting) [reply purchases/src/main/kotlin/com/revenuecat/purchases/errors.kt]: Yeah, in that case it should process like a restore/successful purchase and return a purchaser info, ideally the latest. 

If that's too much of a refactor it can wait until 2.x+.

Comment (vegaro) [reply purchases/src/main/kotlin/com/revenuecat/purchases/errors.kt]: I think I will merge and launch rc1 without it, and launch the stable with that change? I don't want to hold this more :)
